apps/fz_wall/lib/fz_wall/cli/live.ex

```diff
@@ -6,104 +6,139 @@ defmodule FzWall.CLI.Live do
 
   @doc """
-  Deletes nftables rule.
+  Delete rule destination ip from its corresponding sets.
   """
-  def delete_rule(params) do
-    rule_str(params)
-    |> delete_rule_matching()
+  def delete_rule(rule) do
+    remove_from_set(rule.user_id, rule.destination, proto(rule.destination), rule.action)
   end
 
-  def delete_rules(source) do
-    source_match_str(source)
-    |> delete_rule_matching()
+  @doc """
+  Eliminates device rules from its corresponding sets.
+  """
+  def delete_device(device) do
+    get_types()
+    |> Enum.each(fn type -> remove_from_set(device.user_id, device[type], type) end)
   end
 
-  defp delete_rule_matching(rule_str) do
-    rules = exec!("#{nft()} -a list table inet #{@table_name}")
+  defp remove_from_set(_user_id, nil, _type), do: :no_ip
 
-    # When a rule is deleted the others might change handle so we need to
-    # re-scan each time.
-    case rule_handle_regex(~r/^\s*#{rule_str}.*# handle (?<num>\d+)/m, rules) do
-      nil ->
-        :no_rule
+  defp remove_from_set(user_id, ip, type) do
+    get_device_set_name(user_id, type)
+    |> delete_elem(ip)
+  end
 
-      [handle] ->
-        exec!("#{nft()} delete rule inet #{@table_name} forward handle #{handle}")
-        # There might still be matching rules so re-run it
-        delete_rule_matching(rule_str)
-    end
+  defp remove_from_set(user_id, ip, type, action) do
+    get_dest_set_name(user_id, type, action)
+    |> delete_elem(ip)
   end
 
-  @doc """
-  Restores rules.
-  """
-  def restore(rules) do
-    # XXX: Priority?
-    for rule_spec <- rules do
-      add_rule(rule_spec)
-    end
+  defp add_to_set(_user_id, nil, _type), do: :no_ip
+
+  defp add_to_set(user_id, ip, type) do
+    get_device_set_name(user_id, type)
+    |> add_elem(ip)
+  end
+
+  defp add_to_set(user_id, ip, type, action) do
+    get_dest_set_name(user_id, type, action)
+    |> add_elem(ip)
+  end
+
+  defp add_sets(user_id) do
+    list_sets(user_id)
+    |> Enum.each(fn set -> add_set(set) end)
+  end
+
+  defp delete_sets(user_id) do
+    list_sets(user_id)
+    |> Enum.each(fn set -> delete_set(set) end)
+  end
+
+  defp add_rules(user_id) do
+    cross(get_types(), get_actions())
+    |> Enum.each(fn {type, action} ->
+      create_rule(
+        type,
+        get_device_set_name(user_id, type),
+        get_dest_set_name(user_id, type, action),
+        action
+      )
+    end)
+  end
+
+  defp delete_rules(user_id) do
+    cross(get_types(), get_actions())
+    |> Enum.each(fn {type, action} ->
+      remove_rule(
+        type,
+        get_device_set_name(user_id, type),
+        get_dest_set_name(user_id, type, action),
+        action
+      )
+    end)
+  end
+
+  # xxx: here we could add multiple devices/rules in a single nft call
+  def restore({users, devices, rules}) do
+    users |> Enum.each(fn user -> add_user(user) end)
+    devices |> Enum.each(fn device -> add_device(device) end)
+    rules |> Enum.each(fn rule -> add_rule(rule) end)
```

Comment:
Just FYI, because I see a lot of `fn x -> f(x) end` in this PR - it's very common to use the [capture syntax](https://hexdocs.pm/elixir/Kernel.SpecialForms.html#&/1) for simple function calls.

`Enum.each(rules, &add_rule/1)`

Reply:
Oh nice, thanks! I was looking for this.